### PR TITLE
Fix Resolution.Tick TradeBar history request

### PR DESCRIPTION
**Pull request by Martin-Molinero**
#### Description

Based on @gsalaz98 findings exposed on issue https://github.com/QuantConnect/Lean/issues/4435:
- Throw exception if requesting `Resolution.Tick` using the TradeBar
  history request API
- Adding unit tests, asserting current behavior

#### Related Issue

Partially resolution for https://github.com/QuantConnect/Lean/issues/4435
#### Motivation and Context

Improve API user experience
#### Requires Documentation Change

N/A
#### How Has This Been Tested?

Unit and regression tests
#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Algorithm/QCAlgorithm.History.cs [hunk 1]
```diff
@@ -406,6 +400,12 @@ public IEnumerable<TradeBar> History(Symbol symbol, DateTime start, DateTime end
                 Error("Calling History<TradeBar> method on a Forex or CFD security will return an empty result. Please use the generic version with QuoteBar type parameter.");
             }
 
+            if (resolution.HasValue && resolution == Resolution.Tick)
```
Comment (gsalaz98) [inline Algorithm/QCAlgorithm.History.cs:403]: If the `resolution` value is null, will this check be bypassed if the only subscription we have for a given `Symbol` is in `Resolution.Tick`? 

We select the highest resolution available in the added Security if it exists in the following method, which is called from `CreateDateRangeHistoryRequests`:
https://github.com/QuantConnect/Lean/blob/a3c5fd98712be4e8416f08bfb548957e9aabcb4a/Algorithm/QCAlgorithm.History.cs#L707

**Comment by gsalaz98** [comment]
Also, I'm not sure this PR closes the issue #4435 since `History<Tick>(...)` is still only adding the `TickType.Quote` subscription rather than both `TickType.Trade` and `TickType.Quote`, compared to the `History(Symbol[], ...)` call. I've left a comment on the original issue detailing my findings 👍

File: Algorithm/QCAlgorithm.History.cs (hunk 1 above)
Comment (Martin-Molinero) [reply Algorithm/QCAlgorithm.History.cs]: Nice catch! Addressed 👍

**Comment by Martin-Molinero** [comment]
> Also, I'm not sure this PR closes the issue #4435 since `History<Tick>(...)` is still only adding the `TickType.Quote` subscription rather than both `TickType.Trade` and `TickType.Quote`, compared to the `History(Symbol[], ...)` call. I've left a comment on the original issue detailing my findings 👍

True, the difference is that `History(Symbol[], ...)` returns a `Slice` object 🤔. If we were to add both tick types in the result, it would change existing behavior and would end up as a mix of tick types, sometimes Trade, sometimes Quote, somewhat related to issue https://github.com/QuantConnect/Lean/issues/4196.
Another case I though of while reviewing is also related to 4196:

``` CSharp
// Trades and quotes
var result = _algorithm.History(new [] { Symbols.SPY }, start.AddHours(9.8), start.AddHours(10), Resolution.Tick).ToList();
foreach (var slice in result)
{
    if (slice.Ticks.Any(pair => pair.Value.Any(tick => tick.TickType == TickType.Quote) &&
                                pair.Value.Any(tick => tick.TickType == TickType.Trade)))
    {
        // slice.Ticks has both Quote and Trade ticks, what will the following return?
        var ticks = slice.Get<Tick>();
    }
}
```
Let's keep issue #4435 open till we define the expected behavior 👍

**Pull request merged**